File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-0.ts
class Solution {
    canShip(weights: number[], capacityPerShip: number, days: number): boolean {
        let ships: number = 1;
        let currentCapacity = capacityPerShip;

        for (const weight of weights) {
            if (currentCapacity - weight < 0) {
                ships++;
                currentCapacity = capacityPerShip;
            }
            currentCapacity -= weight;
        }
        return ships <= days;
    }

    /**
     * @param {number[]} weights
     * @param {number} days
     * @return {number}
     */
    shipWithinDays(weights: number[], days: number): number {
        let left: number = Math.max(...weights);
        let right: number = 0;
        weights.map((weight: number) => right += weight);
        let res: number = right;

        while (left <= right) {
            let testCapacity: number = Math.floor(left + (right - left) / 2);

            // Check if given this capacity all the weights can be shipped.
            if (this.canShip(weights, testCapacity, days)) {
                res = Math.min(res, testCapacity);
                right = testCapacity - 1;
            } else {
                left = testCapacity + 1;
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/split-array-largest-sum/submission-0.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        left = max(nums)
        right = sum(nums)
        res = right

        while left <= right:
            mid = left + (right - left) // 2

            if self.canSplit(mid, nums, k):
                res = mid
                right = mid - 1
            else:
                left = mid + 1
        
        return res
    
    # largest - The largest sum a given subarray could be.
    def canSplit(self, largest, nums, k):
        subarray = 0
        currentSum = 0

        for n in nums:
            currentSum += n

            if currentSum > largest:
                subarray += 1
                currentSum = n
        
        return subarray + 1 <= k
            
            